### PR TITLE
Bugfix/zcs 2537

### PR DESCRIPTION
**Pull request by tmclane**
Implement AddAccountLogger support in imapd

### CODE REVIEW

**Comment by tmclane** [comment]
Added a commit to fix the 'ValueComparisonTest' failures. Will not include this in the pushed code.

**Comment by tmclane** [comment]
Looks like the 'com.zimbra.cs.filter.RejectTest' tests are failing now and I can't figure out what the issue is.  These test failures are unrelated to my changes but are the reason for the Travis-CI failure.

**Comment by fatkudu** [comment]
@tmclane See https://github.com/Zimbra/zm-mailbox/pull/352 which just fixes ValueComparisonTest

File: store/src/java/com/zimbra/cs/service/admin/AddAccountLogger.java [hunk 1]
```diff
@@ -29,39 +31,38 @@
 import com.zimbra.common.util.LogFactory;
 import com.zimbra.common.util.StringUtil;
 import com.zimbra.common.util.ZimbraLog;
-import com.zimbra.cs.account.Account;
-import com.zimbra.cs.account.AccountServiceException;
-import com.zimbra.cs.account.Provisioning;
+import com.zimbra.cs.account.*;
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/service/admin/AddAccountLogger.java]: Why?

File: store/src/java/com/zimbra/qa/unittest/TestImapViaImapDaemon.java [hunk 2]
```diff
@@ -98,6 +98,36 @@ public void testClearDaemonCacheWrongAuthenticator() throws Exception {
...
+          try {
+              connection.addAccountLogger(account, "zimbra.imap", "trace");
+              fail("should not be able to add an account logger without authenticating");
+          } catch (CommandFailedException cfe) {
+              assertEquals("must be in AUTHENTICATED or SELECTED state", cfe.getError());
+          }
+          connection.login(PASS);
+          try {
+              connection.addAccountLogger(account, "zimbra.imap", "trace");
+              fail("should not be able to  an account logger without using X-ZIMBRA auth mechanism");
```
Comment (fatkudu) [inline store/src/java/com/zimbra/qa/unittest/TestImapViaImapDaemon.java]: missing word "add"

File: store/src/java/com/zimbra/cs/service/admin/AddAccountLogger.java (hunk 1 above)
Comment (tmclane) [reply store/src/java/com/zimbra/cs/service/admin/AddAccountLogger.java]: I believe this is a case of the IDE 'helped' me out and I didn't double check.

File: store/src/java/com/zimbra/cs/imap/ImapHandler.java [hunk 3]
```diff
@@ -46,6 +46,9 @@
 import javax.mail.MessagingException;
 import javax.mail.internet.MimeMessage;
 
+import com.zimbra.common.soap.AdminConstants;
+import com.zimbra.common.soap.Element;
+import com.zimbra.common.util.*;
```
[the remark this answers is not in the document]
Comment (tmclane) [reply store/src/java/com/zimbra/cs/imap/ImapHandler.java]: Fair enough. Will double check the imports and pull in only what is needed.

File: store/src/java/com/zimbra/cs/imap/ImapHandler.java [hunk 4]
```diff
@@ -1485,6 +1491,67 @@ private boolean checkZimbraAdminAuth() {
...
+        ZimbraLog.imap.info("Adding account level logger for '%s' category: %s level %s", id, category, sLevel);
+
+        if (!checkState(tag, State.AUTHENTICATED)) {
+            return true;
+        } else if (!checkZimbraAdminAuth()) {
+            sendNO(tag, "must be authenticated as admin with X-ZIMBRA auth mechanism");
+            return true;
+        }
+
+        try {
```
[the remark this answers is not in the document]
Comment (tmclane) [reply store/src/java/com/zimbra/cs/imap/ImapHandler.java]: I agree.

File: store/src/java/com/zimbra/cs/service/admin/AddAccountLogger.java (hunk 1 above)
Comment (fatkudu) [reply store/src/java/com/zimbra/cs/service/admin/AddAccountLogger.java]: Intellij I'm guessing?

Comment (tmclane) [reply store/src/java/com/zimbra/cs/service/admin/AddAccountLogger.java]: fixed.

File: store/src/java/com/zimbra/qa/unittest/TestImapViaImapDaemon.java (hunk 2 above)
Comment (tmclane) [reply store/src/java/com/zimbra/qa/unittest/TestImapViaImapDaemon.java]: fixed.

File: store/src/java/com/zimbra/cs/imap/ImapHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tmclane) [reply store/src/java/com/zimbra/cs/imap/ImapHandler.java]: Fixed.

File: store/src/java/com/zimbra/cs/imap/ImapHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tmclane) [reply store/src/java/com/zimbra/cs/imap/ImapHandler.java]: Done

**Pull request merged**